hub/subscribe.go

```diff
@@ -17,41 +18,57 @@ func (h *Hub) SubscribeHandler(w http.ResponseWriter, r *http.Request) {
 		panic("The Response Writer must be an instance of Flusher.")
 	}
 
-	subscriber, updateChan, ok := h.initSubscription(w, r)
+	subscriber, pipe, ok := h.initSubscription(w, r)
 	if !ok {
 		return
 	}
 	defer h.cleanup(subscriber)
 
-	for {
-		if h.options.HeartbeatInterval == time.Duration(0) {
+	if h.options.HeartbeatInterval == time.Duration(0) {
+		for {
 			// No heartbeat defined, just block
-			serializedUpdate, open := <-updateChan
-			if !open {
-				return
+			update, err := pipe.Read(context.Background())
+			if err != nil {
+				if err == ErrClosedPipe {
+					return
+				}
+
+				log.Error(err)
+				continue
 			}
-			h.publish(serializedUpdate, subscriber, w, r)
 
-			continue
+			h.publish(newSerializedUpdate(update), subscriber, w, r)
 		}
+	}
 
-		select {
-		case serializedUpdate, open := <-updateChan:
-			if !open {
+	for {
+		ctx, cancel := context.WithTimeout(context.Background(), h.options.HeartbeatInterval)
+		update, err := pipe.Read(ctx)
+		cancel()
+
+		if err != nil {
+			if err == ErrClosedPipe {
 				return
 			}
-			h.publish(serializedUpdate, subscriber, w, r)
 
-		case <-time.After(h.options.HeartbeatInterval):
-			// Send a SSE comment as a heartbeat, to prevent issues with some proxies and old browsers
-			fmt.Fprint(w, ":\n")
-			f.Flush()
+			if err == context.DeadlineExceeded {
+				// Send a SSE comment as a heartbeat, to prevent issues with some proxies and old browsers
+				fmt.Fprint(w, ":\n")
+				f.Flush()
+
+				continue
+			}
+
+			log.Error(err)
```

Comment:
This line isn't tested, it could be tested using a cancellable context in the test.